### ISSUE TITLE
Bug: Poor performance for `HypreBoomerAMG` with `Ordering::byNODES`

### ISSUE THREAD

**Issue by zatkins-work** [mfem/mfem#4404]
# Bug Report
If `HypreBoomerAMG` is used with `Ordering::byNODES`, then the number of PCG iterations needed for a simple elasticity problem is increased by a factor of 3-5. For comparison, I used the `examples/ex2p.cpp` example. 

At least for this example, the performance degradation is only seen when using the `HyperBoomerAMG::SetElasticityOptions` function. However, this bug was originally noted in Serac, where the difference is present when using either the  `HyperBoomerAMG::SetElasticityOptions` or `HyperBoomerAMG::SetSystemsOptions` functions (see [LLNL/serac#1175](https://github.com/LLNL/serac/issues/1175)).

## `Ordering::byVDIM` results
Run options:
```console
mpirun -np 4 ex2p -m ../data/beam-hex.mesh -elast -vdim
```

The number of PCG iterations is 28 and the wall clock time is 0.193745 seconds. Full output is given below

<details>

```
Options used:
   --mesh mfem/data/beam-hex.mesh
   --order 1
   --amg-for-elasticity
   --no-static-condensation
   --visualization
   --by-vdim
   --device cpu
Device configuration: cpu
Memory configuration: host-std
Number of finite element unknowns: 15795
Assembling: r.h.s. ... matrix ... done.
Size of linear system: 15795


 Num MPI tasks = 4

 Num OpenMP threads = 1


BoomerAMG SETUP PARAMETERS:

 Max levels = 25
 Num levels = 7

 Strength Threshold = 0.500000
 Interpolation Truncation Factor = 0.000000
 Maximum Row Sum Threshold for Dependency Weakening = 0.900000

 Coarsening Type = HMIS 
 measures are determined locally


 No global partition option chosen.

 Interpolation = extended+i interpolation

Operator Matrix Information:

             nonzero            entries/row          row sums
lev    rows  entries sparse   min  max     avg      min         max
======================================================================
  0   15795  1085625  0.004    24   81    68.7  -2.105e-14   1.875e+01
  1    7824  2486036  0.041    92  624   317.7  -6.756e+00   2.893e+01
  2    1890   680308  0.190    96  552   360.0  -9.728e+00   3.449e+01
  3     288    38096  0.459    58  201   132.3  -3.187e+01   6.649e+01
  4      30      900  1.000    30   30    30.0  -4.937e+01   1.008e+02
  5      12      144  1.000    12   12    12.0  -1.147e+01   1.031e+02
  6       6       36  1.000     6    6     6.0  -1.183e+01   1.035e+02


Interpolation Matrix Information:
                    entries/row        min        max            row sums
lev  rows x cols  min  max  avgW     weight      weight       min         max
================================================================================
  0 15795 x 7824    0   12  15.5  -7.156e-01   1.063e+00  -4.312e-01   1.250e+00
  1  7824 x 1890    0   12   7.7  -1.428e+00   2.424e+00  -4.327e-01   3.240e+00
  2  1890 x 288     0   12   7.5  -3.867e+00   4.431e+00  -4.979e-01   2.105e+00
  3   288 x 30      0   12   4.0  -9.830e-01   1.000e+00  -3.900e-01   2.370e+00
  4    30 x 12      0    6   2.2  -2.992e+00   1.646e+00  -1.976e+00   2.998e+00
  5    12 x 6       0    3   1.7  -3.551e+00   3.530e+00  -2.057e+00   3.770e+00


     Complexity:    grid = 1.636277
                operator = 3.952695
                memory = 4.130289




BoomerAMG SOLVER PARAMETERS:

  Maximum number of cycles:         1 
  Stopping Tolerance:               0.000000e+00 
  Cycle type (1 = V, 2 = W, etc.):  1

  Relaxation Parameters:
   Visiting Grid:                     down   up  coarse
            Number of sweeps:            1    1     1 
   Type 0=Jac, 3=hGS, 6=hSGS, 9=GE:      8    8     8 
   Point types, partial sweeps (1=C, -1=F):
                  Pre-CG relaxation (down):   0
                   Post-CG relaxation (up):   0
                             Coarsest grid:   0

=============================================
Setup phase times:
=============================================
PCG Setup:
  wall clock time = 0.110436 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.109974 seconds
  cpu MFLOPS      = 0.000000

<C*b,b>: 4.770840e-04


Iters       ||r||_C     conv.rate  ||r||_C/||b||_C
-----    ------------    ---------  ------------ 
    1    3.928342e-02    1.798506    1.798506e+00
    2    3.010079e-02    0.766247    1.378099e+00
    3    1.479262e-02    0.491436    6.772481e-01
    4    6.511921e-03    0.440214    2.981341e-01
    5    2.755099e-03    0.423085    1.261362e-01
    6    2.396935e-03    0.870000    1.097385e-01
    7    1.987926e-03    0.829362    9.101288e-02
    8    1.033667e-03    0.519972    4.732417e-02
    9    6.627327e-04    0.641147    3.034177e-02
   10    3.736817e-04    0.563850    1.710820e-02
   11    2.503621e-04    0.669988    1.146229e-02
   12    1.727431e-04    0.689973    7.908669e-03
   13    2.117631e-04    1.225884    9.695112e-03
   14    1.144711e-04    0.540562    5.240811e-03
   15    4.477617e-05    0.391157    2.049980e-03
   16    3.050331e-05    0.681240    1.396528e-03
   17    1.775814e-05    0.582171    8.130179e-04
   18    1.806086e-05    1.017047    8.268771e-04
   19    7.589884e-06    0.420239    3.474863e-04
   20    3.800416e-06    0.500721    1.739938e-04
   21    1.302926e-06    0.342838    5.965161e-05
   22    4.135177e-07    0.317376    1.893201e-05
   23    1.205287e-07    0.291472    5.518146e-06
   24    4.649062e-08    0.385722    2.128472e-06
   25    1.367644e-08    0.294176    6.261460e-07
   26    2.758539e-09    0.201700    1.262937e-07
   27    6.849848e-10    0.248314    3.136054e-08
   28    1.030533e-10    0.150446    4.718072e-09


=============================================
Solve phase times:
=============================================
PCG Solve:
  wall clock time = 0.193745 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.190707 seconds
  cpu MFLOPS      = 0.000000

PCG Iterations = 28
Final PCG Relative Residual Norm = 4.71807e-09
```

</details>

## `Ordering::byNODES` results
Run options:
```console
mpirun -np 4 ex2p -m ../data/beam-hex.mesh -elast -nodes
```

The number of PCG iterations is 162 and the wall clock time is 0.616807 seconds. Full output is given below

<details>

```
Options used:
   --mesh mfem/data/beam-hex.mesh
   --order 1
   --amg-for-elasticity
   --no-static-condensation
   --visualization
   --by-nodes
   --device cpu
Device configuration: cpu
Memory configuration: host-std
Number of finite element unknowns: 15795
Assembling: r.h.s. ... matrix ... done.
Size of linear system: 15795


 Num MPI tasks = 4

 Num OpenMP threads = 1


BoomerAMG SETUP PARAMETERS:

 Max levels = 25
 Num levels = 7

 Strength Threshold = 0.500000
 Interpolation Truncation Factor = 0.000000
 Maximum Row Sum Threshold for Dependency Weakening = 0.900000

 Coarsening Type = HMIS 
 measures are determined locally


 No global partition option chosen.

 Interpolation = extended+i interpolation

Operator Matrix Information:

             nonzero            entries/row          row sums
lev    rows  entries sparse   min  max     avg      min         max
======================================================================
  0   15795  1085625  0.004    24   81    68.7  -2.105e-14   1.875e+01
  1    7656  1300680  0.022     1  588   169.9  -1.047e+02   5.043e+02
  2    1824   295154  0.089     1  390   161.8  -5.542e+29   2.940e+31
  3     312    23514  0.242    11  142    75.4  -4.945e+29   2.940e+31
  4      48      868  0.377     9   24    18.1  -1.176e+16   2.939e+31
  5      12      108  0.750     6   12     9.0  -1.176e+16   2.939e+31
  6       6       36  1.000     6    6     6.0  -1.176e+16   2.939e+31


Interpolation Matrix Information:
                    entries/row        min        max            row sums
lev  rows x cols  min  max  avgW     weight      weight       min         max
================================================================================
  0 15795 x 7656    0   15   4.6  -9.000e+00   7.500e+00  -1.187e+01   1.575e+01
  1  7656 x 1824    0   16   1.7  -1.293e+17   3.485e+17  -1.293e+17   3.485e+17
  2  1824 x 312     0   12   1.2  -2.627e+13   3.431e+15  -2.627e+13   3.431e+15
  3   312 x 48      0    7   1.0  -6.378e+00   9.215e+10  -6.000e+00   9.215e+10
  4    48 x 12      0    4   1.1  -7.375e+00   2.875e+00  -7.250e+00   2.000e+00
  5    12 x 6       0    1   0.2   2.417e-35   2.417e-35   0.000e+00   1.000e+00


     Complexity:    grid = 1.624122
                operator = 2.492560
                memory = 2.546757




BoomerAMG SOLVER PARAMETERS:

  Maximum number of cycles:         1 
  Stopping Tolerance:               0.000000e+00 
  Cycle type (1 = V, 2 = W, etc.):  1

  Relaxation Parameters:
   Visiting Grid:                     down   up  coarse
            Number of sweeps:            1    1     1 
   Type 0=Jac, 3=hGS, 6=hSGS, 9=GE:      8    8     8 
   Point types, partial sweeps (1=C, -1=F):
                  Pre-CG relaxation (down):   0
                   Post-CG relaxation (up):   0
                             Coarsest grid:   0

=============================================
Setup phase times:
=============================================
PCG Setup:
  wall clock time = 0.046780 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.045841 seconds
  cpu MFLOPS      = 0.000000

<C*b,b>: 1.635386e-05


Iters       ||r||_C     conv.rate  ||r||_C/||b||_C
-----    ------------    ---------  ------------ 
    1    3.427763e-03    0.847619    8.476190e-01
    2    4.626211e-03    1.349630    1.143972e+00
    3    5.819511e-03    1.257943    1.439052e+00
    4    5.672679e-03    0.974769    1.402743e+00
    5    5.622712e-03    0.991192    1.390387e+00
    6    6.761773e-03    1.202582    1.672055e+00
    7    6.934473e-03    1.025541    1.714760e+00
    8    6.905494e-03    0.995821    1.707594e+00
    9    7.413967e-03    1.073633    1.833330e+00
   10    7.167698e-03    0.966783    1.772432e+00
   11    7.591442e-03    1.059119    1.877216e+00
   12    7.301626e-03    0.961823    1.805550e+00
   13    7.168101e-03    0.981713    1.772532e+00
   14    7.296312e-03    1.017886    1.804236e+00
   15    5.913727e-03    0.810509    1.462350e+00
   16    4.375678e-03    0.739919    1.082020e+00
   17    6.292031e-03    1.437956    1.555897e+00
   18    1.100648e-02    1.749273    2.721688e+00
   19    1.618628e-02    1.470613    4.002551e+00
   20    1.735969e-02    1.072494    4.292712e+00
   21    1.295650e-02    0.746356    3.203891e+00
   22    1.259144e-02    0.971824    3.113617e+00
   23    1.316006e-02    1.045160    3.254227e+00
   24    1.096962e-02    0.833554    2.712574e+00
   25    8.706963e-03    0.793734    2.153062e+00
   26    8.571999e-03    0.984499    2.119688e+00
   27    7.642967e-03    0.891620    1.889957e+00
   28    6.801277e-03    0.889874    1.681823e+00
   29    6.613880e-03    0.972447    1.635484e+00
   30    5.301978e-03    0.801644    1.311076e+00
   31    4.392970e-03    0.828553    1.086296e+00
   32    4.497335e-03    1.023757    1.112103e+00
   33    4.861421e-03    1.080956    1.202135e+00
   34    6.354051e-03    1.307036    1.571233e+00
   35    8.445002e-03    1.329074    2.088284e+00
   36    7.559789e-03    0.895179    1.869389e+00
   37    3.766719e-03    0.498257    9.314361e-01
   38    1.825633e-03    0.484675    4.514435e-01
   39    1.166833e-03    0.639139    2.885350e-01
   40    1.160023e-03    0.994164    2.868511e-01
   41    1.215876e-03    1.048148    3.006624e-01
   42    1.248594e-03    1.026909    3.087529e-01
   43    1.296311e-03    1.038216    3.205524e-01
   44    1.754694e-03    1.353606    4.339016e-01
   45    3.209816e-03    1.829274    7.937249e-01
   46    5.515813e-03    1.718420    1.363953e+00
   47    6.522968e-03    1.182594    1.613003e+00
   48    5.306013e-03    0.813435    1.312074e+00
   49    4.557954e-03    0.859017    1.127093e+00
   50    4.639401e-03    1.017869    1.147233e+00
   51    4.708707e-03    1.014939    1.164372e+00
   52    4.453323e-03    0.945763    1.101220e+00
   53    5.020622e-03    1.127388    1.241502e+00
   54    4.967291e-03    0.989378    1.228314e+00
   55    3.749726e-03    0.754883    9.272341e-01
   56    3.285491e-03    0.876195    8.124378e-01
   57    3.189066e-03    0.970651    7.885938e-01
   58    3.179360e-03    0.996957    7.861937e-01
   59    2.822002e-03    0.887601    6.978262e-01
   60    2.538912e-03    0.899685    6.278235e-01
   61    2.670893e-03    1.051984    6.604599e-01
   62    3.055873e-03    1.144139    7.556579e-01
   63    2.767548e-03    0.905649    6.843608e-01
   64    1.657906e-03    0.599052    4.099678e-01
   65    8.892601e-04    0.536376    2.198967e-01
   66    6.569345e-04    0.738743    1.624471e-01
   67    7.294546e-04    1.110392    1.803799e-01
   68    9.405830e-04    1.289433    2.325879e-01
   69    8.539592e-04    0.907904    2.111675e-01
   70    6.139685e-04    0.718967    1.518224e-01
   71    5.319762e-04    0.866455    1.315474e-01
   72    5.705848e-04    1.072576    1.410945e-01
   73    6.476844e-04    1.135124    1.601597e-01
   74    6.444918e-04    0.995071    1.593703e-01
   75    5.872131e-04    0.911126    1.452064e-01
   76    5.201482e-04    0.885791    1.286225e-01
   77    5.007679e-04    0.962741    1.238301e-01
   78    4.713052e-04    0.941165    1.165446e-01
   79    3.183220e-04    0.675405    7.871482e-02
   80    2.113566e-04    0.663971    5.226436e-02
   81    2.064717e-04    0.976888    5.105644e-02
   82    2.701979e-04    1.308644    6.681467e-02
   83    3.382492e-04    1.251857    8.364244e-02
   84    3.622039e-04    1.070820    8.956597e-02
   85    3.063490e-04    0.845792    7.575414e-02
   86    2.353704e-04    0.768308    5.820252e-02
   87    1.826247e-04    0.775903    4.515953e-02
   88    1.903772e-04    1.042450    4.707657e-02
   89    1.944019e-04    1.021141    4.807180e-02
   90    1.939195e-04    0.997518    4.795250e-02
   91    1.604543e-04    0.827428    3.967723e-02
   92    9.151549e-05    0.570352    2.263000e-02
   93    5.843188e-05    0.638492    1.444907e-02
   94    4.041724e-05    0.691698    9.994397e-03
   95    3.714297e-05    0.918988    9.184734e-03
   96    3.203927e-05    0.862593    7.922688e-03
   97    2.805461e-05    0.875632    6.937359e-03
   98    2.603854e-05    0.928137    6.438822e-03
   99    3.328793e-05    1.278410    8.231456e-03
  100    4.872824e-05    1.463841    1.204954e-02
  101    6.544331e-05    1.343026    1.618286e-02
  102    5.482430e-05    0.837737    1.355698e-02
  103    2.652137e-05    0.483752    6.558218e-03
  104    1.205323e-05    0.454472    2.980529e-03
  105    6.146482e-06    0.509945    1.519905e-03
  106    4.073934e-06    0.662807    1.007404e-03
  107    2.982548e-06    0.732105    7.375261e-04
  108    2.995564e-06    1.004364    7.407446e-04
  109    2.927599e-06    0.977312    7.239383e-04
  110    2.369314e-06    0.809303    5.858852e-04
  111    1.969558e-06    0.831278    4.870332e-04
  112    1.995850e-06    1.013349    4.935347e-04
  113    1.835586e-06    0.919701    4.539046e-04
  114    1.851376e-06    1.008602    4.578093e-04
  115    2.148967e-06    1.160740    5.313977e-04
  116    2.177935e-06    1.013480    5.385610e-04
  117    2.402639e-06    1.103173    5.941259e-04
  118    3.130981e-06    1.303143    7.742307e-04
  119    4.635370e-06    1.480485    1.146237e-03
  120    4.691439e-06    1.012096    1.160101e-03
  121    3.022072e-06    0.644167    7.472994e-04
  122    1.968967e-06    0.651529    4.868871e-04
  123    1.498696e-06    0.761158    3.705982e-04
  124    1.751585e-06    1.168740    4.331329e-04
  125    2.927771e-06    1.671498    7.239806e-04
  126    4.867678e-06    1.662589    1.203682e-03
  127    5.344851e-06    1.098029    1.321678e-03
  128    5.256341e-06    0.983440    1.299791e-03
  129    6.168965e-06    1.173623    1.525465e-03
  130    6.839685e-06    1.108725    1.691321e-03
  131    4.703117e-06    0.687622    1.162989e-03
  132    2.463254e-06    0.523749    6.091148e-04
  133    1.417473e-06    0.575447    3.505134e-04
  134    1.089488e-06    0.768613    2.694091e-04
  135    1.056142e-06    0.969394    2.611635e-04
  136    1.333225e-06    1.262353    3.296805e-04
  137    1.776750e-06    1.332671    4.393557e-04
  138    2.075086e-06    1.167911    5.131284e-04
  139    1.961597e-06    0.945309    4.850648e-04
  140    1.208893e-06    0.616280    2.989357e-04
  141    5.963222e-07    0.493280    1.474589e-04
  142    2.781944e-07    0.466517    6.879205e-05
  143    2.029210e-07    0.729421    5.017840e-05
  144    2.819280e-07    1.389349    6.971529e-05
  145    5.539044e-07    1.964702    1.369698e-04
  146    1.157805e-06    2.090262    2.863027e-04
  147    2.349269e-06    2.029071    5.809284e-04
  148    3.406379e-06    1.449974    8.423312e-04
  149    2.728815e-06    0.801090    6.747828e-04
  150    1.329768e-06    0.487306    3.288257e-04
  151    5.387322e-07    0.405133    1.332180e-04
  152    1.978795e-07    0.367306    4.893175e-05
  153    7.575288e-08    0.382823    1.873221e-05
  154    3.174400e-08    0.419047    7.849672e-06
  155    1.396222e-08    0.439838    3.452585e-06
  156    6.781540e-09    0.485706    1.676943e-06
  157    3.237426e-09    0.477388    8.005523e-07
  158    1.233442e-09    0.380995    3.050062e-07
  159    4.551153e-10    0.368980    1.125412e-07
  160    1.713708e-10    0.376544    4.237667e-08
  161    6.869889e-11    0.400879    1.698790e-08
  162    3.019790e-11    0.439569    7.467352e-09


=============================================
Solve phase times:
=============================================
PCG Solve:
  wall clock time = 0.616807 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.607428 seconds
  cpu MFLOPS      = 0.000000

PCG Iterations = 162
Final PCG Relative Residual Norm = 7.46735e-09
```

</details>

# Possible Causes
The most likely issue is that the permutation/DOF function defined in `HyperBoomerAMG::SetSystemsOptions` is wrong, leading to incorrectly built rigid-body mode vectors.

Alternatively, the permutation is correct, but is not propagated to coarsened meshes.

**Comment by tzanio** [comment]
What happens without `-elast`?

**Comment by jandrej** [comment]
If you use the `-elast` option, `SetSystemsOptions` is ignored.

**Comment by tzanio** [comment]
I wonder if the elasticity solver (with option `-elast`) works only with `Ordering::byVDIM` and the changes in this PR need to be augmented: https://github.com/mfem/mfem/pull/3772/files

**Comment by justinlaughlin** [comment]
I was curious and tried comparing `-nodes` vs `-vdim` without `-elast` and it is about the same # of cg iterations. 

```mpirun -np 4 ./ex2p -m ../data/beam-hex.mesh -vdim```
<details>

```console
=============================================
Solve phase times:
=============================================
PCG Solve:
  wall clock time = 0.177536 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.177494 seconds
  cpu MFLOPS      = 0.000000

PCG Iterations = 35
Final PCG Relative Residual Norm = 8.63456e-09
```

</details>

```mpirun -np 4 ./ex2p -m ../data/beam-hex.mesh -nodes```

<details>

```console
=============================================
Solve phase times:
=============================================
PCG Solve:
  wall clock time = 0.152169 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.152169 seconds
  cpu MFLOPS      = 0.000000

PCG Iterations = 36
Final PCG Relative Residual Norm = 1.79992e-09
```

</details>

Regarding the slack thread about seeing differences in performance between `bynodes` vs `byvdim`, both when called using `SetSystemOptions`, it seems like in this example `-elast`  also [changes the fespace constructor](https://github.com/mfem/mfem/blob/master/examples/ex2p.cpp#L159-L176). Maybe that is another thing to check?

**Comment by zatkins-work** [comment]
> Regarding the slack thread about seeing differences in performance between `bynodes` vs `byvdim`, both when called using `SetSystemOptions`, it seems like in this example `-elast` also [changes the fespace constructor](https://github.com/mfem/mfem/blob/master/examples/ex2p.cpp#L159-L176). Maybe that is another thing to check?

I think that it only changes the constructor if we are using a NURBS mesh (which the `beam-hex.mesh` mesh isn't. Am I missing something?

**Comment by justinlaughlin** [comment]
You're right, I could have worded that better. My point is: in this example, when the options are called using `SetSystemOptions` the performance is similar. If you are seeing differences in serac and also using `SetSystemOptions`, my hunch would be to look at the constructor (in this example, the constructor is also changed when using the `nodes` vs `vdim` flag).

**Comment by tzanio** [comment]
I need to dig in the hypre code to provide more details, but I believe the `-elast -nodes` option is not really valid and probably shouldn't be allowed (that was the behavior before https://github.com/mfem/mfem/pull/3772/files).

The three options that make sense to me are:

#### 1. System AMG with ordering by VDIM (default)

```
mpirun -np 4 ex2p -m ../data/beam-hex.mesh
```

or 

```
mpirun -np 4 ex2p -m ../data/beam-hex.mesh -vdim
```


#### 2. System AMG with ordering by NODES

```
mpirun -np 4 ex2p -m ../data/beam-hex.mesh -nodes
```


#### 3. Elasticity-AMG solver, implying and only working for ordering by VDIM

```
mpirun -np 4 ex2p -m ../data/beam-hex.mesh -elast
```

I would expect options 1 and 2 to have similar performance and option 3 to be more expensive and setup and iteration time, but potentially requiring fewer iterations (depending on the problem).

Any comments or additional suggestions?

cc: @ulrikeyang, @liruipeng

**Comment by tzanio** [comment]
Looking more carefully at the *hypre* sources

- https://github.com/hypre-space/hypre/blob/master/src/parcsr_ls/par_sv_interp.c#L197
- https://github.com/hypre-space/hypre/blob/master/src/parcsr_ls/par_sv_interp_ln.c#L27

and this comment in https://doi.org/10.1002/nla.688:

> We refer to the combination of a node-based coarsening with an unknown-based interpolation as a ‘hybrid’ approach. The two new methods for extending AMG interpolation operators that wepresent in Sections 4 and 5 are based on this hybrid approach.

I think it is safe to assume that the "Elasticity" version of BoomerAMG (enabled in MFEM with `SetElasticityOptions`) does require a nodal ordering of the matrix, which in MFEM means `Ordering::byVDIM`.

I suggest that we revert https://github.com/mfem/mfem/pull/3772 except for [this change](https://github.com/mfem/mfem/pull/3772/files#diff-aca2cc45cb0c576bb64a3e0d5c9364f2e5fb75d21281fee73670c230bb66a221L5104-R5105) and instead give a warning if the ordering of the provided space is not `byVDIM`.

We can also perform a similar check in Example 2p.

@zatkins-work, @samuelpmishLLNL and @jamiebramwell -- will the above change be acceptable for you?

**Comment by samuelpmishLLNL** [comment]
> @zatkins-work, @samuelpmishLLNL and @jamiebramwell -- will the above change be acceptable for you?

Sounds good to me. 

> I suggest that we ... give a warning if the ordering of the provided space is not byVDIM.

:+1:  for the warning, but let's also document this requirement in the Doxygen entry for that function.

**Comment by zatkins-work** [comment]
I agree with @samuelpmishLLNL

**Comment by tzanio** [comment]
Fix proposed in #4412, please take a look and comment there.
